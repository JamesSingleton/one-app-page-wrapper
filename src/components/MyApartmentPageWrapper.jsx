import React from 'react';
import { connectAsync } from 'iguazu';
import { queryModule } from 'iguazu-holocron';
import {
  shape, func, string, node,
} from 'prop-types';

const MyApartmentPageWrapper = ({
  loadStatus, Header, Footer, children,
}) => {
  if (loadStatus.all === 'loading') {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Header />
      {children}
      <Footer />
    </div>
  );
};

MyApartmentPageWrapper.propTypes = {
  loadStatus: shape({
    all: string,
  }).isRequired,
  Header: func.isRequired,
  Footer: func.isRequired,
  children: node.isRequired,
};

const loadDataAsProps = ({ store: { dispatch } }) => ({
  Header: () => dispatch(queryModule('my-apartment-header')),
  Footer: () => dispatch(queryModule('my-apartment-footer')),
});

export default connectAsync({ loadDataAsProps })(MyApartmentPageWrapper);
