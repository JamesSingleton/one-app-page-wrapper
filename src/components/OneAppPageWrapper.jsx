import React from 'react';
import { connectAsync } from 'iguazu';
import { queryModule } from 'iguazu-holocron';
import {
  shape, func, string, node,
} from 'prop-types';

const OneAppPageWrapper = ({
  loadStatus, Header, Footer, children,
}) => {
  if (loadStatus.all === 'loading') {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <Header />
      {children}
      <Footer />
    </div>
  );
};

OneAppPageWrapper.propTypes = {
  loadStatus: shape({
    all: string,
  }).isRequired,
  Header: func.isRequired,
  Footer: func.isRequired,
  children: node.isRequired,
};

const loadDataAsProps = ({ store: { dispatch }, ownProps }) => {
  const { header, footer } = ownProps.route;

  return {
    Header: () => dispatch(queryModule(header)),
    Footer: () => dispatch(queryModule(footer)),
  };
};

export default connectAsync({ loadDataAsProps })(OneAppPageWrapper);
