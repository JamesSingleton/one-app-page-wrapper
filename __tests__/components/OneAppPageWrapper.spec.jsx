import React from 'react';
import { shallow } from 'enzyme';

import { ComponentUnderTest as OneAppPageWrapper, loadDataAsProps } from '../../src/components/OneAppPageWrapper';

jest.mock('iguazu-holocron', () => ({
  queryModule: jest.fn((moduleName) => `queryModule result for ${moduleName}`),
}));

const Header = () => <div>Mock Header</div>;
const Footer = () => <div>Mock Footer</div>;
const MockChild = () => <div>Mock Child</div>;

describe('OneAppPageWrapper', () => {
  const { NODE_ENV } = process.env;

  afterEach(() => {
    process.env.NODE_ENV = NODE_ENV;
  });

  it('should render the correct loading', () => {
    const props = {
      loadStatus: {
        all: 'loading',
      },
      Header,
      Footer,
      children: <MockChild />,
    };
    const renderedModule = shallow(<OneAppPageWrapper {...props} />);
    expect(renderedModule).toMatchSnapshot();
  });

  it('should render the correct loaded state', () => {
    const props = {
      loadStatus: {
        all: 'complete',
      },
      Header,
      Footer,
      children: <MockChild />,
    };
    const renderedModule = shallow(<OneAppPageWrapper {...props} />);
    expect(renderedModule).toMatchSnapshot();
  });
});

describe('loadDataAsProps', () => {
  const dispatch = jest.fn((x) => x);
  const ownProps = {
    route: {
      header: <Header />,
      footer: <Footer />,
    },
  };
  const loadFuncMap = loadDataAsProps({ store: { dispatch }, ownProps });

  beforeEach(() => dispatch.mockClear());

  it('should load the Header', () => {
    const headerAsyncState = loadFuncMap.Header();
    expect(dispatch).toHaveBeenCalledWith(headerAsyncState);
  });

  it('should load the Footer', () => {
    const footerAsyncState = loadFuncMap.Footer();
    expect(dispatch).toHaveBeenCalledWith(footerAsyncState);
  });
});
