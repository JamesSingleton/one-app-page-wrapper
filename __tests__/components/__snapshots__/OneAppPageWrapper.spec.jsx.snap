// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneAppPageWrapper should add classname 1`] = `
<div
  className="my-class-name"
>
  <Header />
  <MockChild />
  <Footer />
</div>
`;

exports[`OneAppPageWrapper should render the correct loaded state 1`] = `
<div
  className=""
>
  <Header />
  <MockChild />
  <Footer />
</div>
`;

exports[`OneAppPageWrapper should render the correct loading 1`] = `
<div>
  Loading...
</div>
`;
